### ISSUE TITLE
Reference to undefined settings error during bloopInstall with Lagom plugin

### ISSUE THREAD

**Issue by matthewpflueger** [scalacenter/bloop#429]
This seems like a conflict with the Lagom plugin and its internal projects. I have a fairly simple project using Lagom 1.4.4 and receive this error when running bloopInstall (I'm using bloop M9):

```
[error] sbt.internal.util.Init$RuntimeUndefined: References to undefined settings at runtime.
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-kafka)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-cassandra)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-kafka)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-service-locator)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-service-locator)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-cassandra)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error]         at sbt.internal.util.Init.sbt$internal$util$Init$$handleUndefined(Settings.scala:680)
[error]         at sbt.internal.util.Init$Bind.$anonfun$validateKeyReferenced$3(Settings.scala:772)
[error]         at sbt.internal.util.Init$Bind.evaluate(Settings.scala:767)
[error]         at sbt.std.FullInstance$.$anonfun$flatten$2(TaskMacro.scala:68)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:39)
[error]         at sbt.std.Transform$$anon$4.work(System.scala:66)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:262)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error]         at sbt.Execute.work(Execute.scala:271)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:262)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:174)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:36)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error]         at java.lang.Thread.run(Thread.java:748)
[error] (Global / bloopInstall) sbt.internal.util.Init$RuntimeUndefined: References to undefined settings at runtime.
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-kafka)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-cassandra)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-kafka)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-service-locator)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-service-locator)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/vp/scm/,lagom-internal-meta-project-cassandra)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
```

**Comment by jvican** [comment]
So you just followed the instructions in the README, and you got this error?
  
Are you or the lagom plugin using source dependencies? This error can only happen if:
  
1. The plugin is not applied to these projects (most likely).
1. Sbt or a plugin is nuking the project settings coming from Bloop (less likely).
  
Can you copy paste here the output of the `plugins` and `projects` sbt commands?

**Comment by matthewpflueger** [comment]
Sorry for the delayed response! Yes, after following the instructions exactly I received the error and nobody is using source dependencies. The steps to repeat the issue with a fresh repo (https://github.com/matthewpflueger/scalacenter_bloop_issues_429) are below:

1. Follow the instructions [here](https://www.lagomframework.com/documentation/1.4.x/scala/IntroGetStarted.html) to create the example hello Lagom project (which basically amounts to running sbt -Dsbt.version=0.13.16 new https://github.com/lagom/lagom-scala.g8 && sbt runAll)
2. Verify that the hello example is running properly (navigate to http://localhost:9000/api/hello/World)
3. Follow the instructions [here](https://scalacenter.github.io/bloop/docs/installation/) to install bloop on MacOs and ensure it is working (basically brew install scalacenter/bloop/bloop && bloop about)
4. Stopped the hello example, exited out of sbt, added addSbtPlugin("ch.epfl.scala" % "sbt-bloop" % "1.0.0-M9") to the project/plugins.sbt
5. run sbt bloopInstall

### Full output of sbt bloopInstall
```
[/opt/hello]$ sbt bloopInstall
[info] Loading settings from plugins.sbt ...
[info] Loading project definition from /opt/hello/project
[info] Updating ProjectRef(uri("file:/opt/hello/project/"), "hello-build")...
[info] Done updating.
[warn] Found version conflict(s) in library dependencies; some are suspected to be binary incompatible:
[warn]  * org.webjars:webjars-locator-core:0.33 is selected over 0.32
[warn]      +- com.typesafe:npm_2.12:1.2.1                        (depends on 0.32)
[warn]      +- com.typesafe.sbt:sbt-web:1.4.3 (scalaVersion=2.12, sbtVersion=1.0) (depends on 0.32)
[warn]  * org.codehaus.plexus:plexus-utils:3.0.17 is selected over {2.1, 1.5.5}
[warn]      +- org.apache.maven:maven-settings:3.2.2              (depends on 3.0.17)
[warn]      +- org.apache.maven:maven-repository-metadata:3.2.2   (depends on 3.0.17)
[warn]      +- org.apache.maven:maven-aether-provider:3.2.2       (depends on 3.0.17)
[warn]      +- org.apache.maven:maven-model:3.2.2                 (depends on 3.0.17)
[warn]      +- org.apache.maven:maven-core:3.2.2                  (depends on 3.0.17)
[warn]      +- org.apache.maven:maven-artifact:3.2.2              (depends on 3.0.17)
[warn]      +- org.apache.maven:maven-settings-builder:3.2.2      (depends on 3.0.17)
[warn]      +- org.apache.maven:maven-model-builder:3.2.2         (depends on 3.0.17)
[warn]      +- org.sonatype.plexus:plexus-sec-dispatcher:1.3      (depends on 1.5.5)
[warn]      +- org.eclipse.sisu:org.eclipse.sisu.plexus:0.0.0.M5  (depends on 2.1)
[warn]  * com.github.jnr:jnr-constants:0.9.0 is selected over 0.8.7
[warn]      +- com.github.jnr:jnr-posix:3.0.27                    (depends on 0.9.0)
[warn]      +- com.github.jnr:jnr-unixsocket:0.8                  (depends on 0.8.7)
[warn]      +- com.github.jnr:jnr-enxio:0.9                       (depends on 0.8.7)
[warn]  * com.google.guava:guava:23.0 is selected over {10.0.1, 16.0, 20.0, 19.0}
[warn]      +- io.methvin:directory-watcher:0.3.2                 (depends on 23.0)
[warn]      +- com.datastax.cassandra:cassandra-driver-core:3.2.0 (depends on 19.0)
[warn]      +- com.fasterxml.jackson.datatype:jackson-datatype-guava:2.8.8 (depends on 10.0.1)
[warn]      +- org.eclipse.sisu:org.eclipse.sisu.plexus:0.0.0.M5  (depends on 10.0.1)
[warn]      +- com.spotify:docker-client:8.9.0                    (depends on 10.0.1)
[warn] Run 'evicted' to see detailed eviction warnings
[info] Loading settings from build.sbt ...
[info] Set current project to hello (in build file:/opt/hello/)
[error] sbt.internal.util.Init$RuntimeUndefined: References to undefined settings at runtime.
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-cassandra)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-service-locator)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-kafka)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-service-locator)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-cassandra)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-kafka)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error]         at sbt.internal.util.Init.sbt$internal$util$Init$$handleUndefined(Settings.scala:680)
[error]         at sbt.internal.util.Init$Bind.$anonfun$validateKeyReferenced$3(Settings.scala:772)
[error]         at sbt.internal.util.Init$Bind.evaluate(Settings.scala:767)
[error]         at sbt.std.FullInstance$.$anonfun$flatten$2(TaskMacro.scala:68)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:39)
[error]         at sbt.std.Transform$$anon$4.work(System.scala:66)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:262)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error]         at sbt.Execute.work(Execute.scala:271)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:262)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:174)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:36)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error]         at java.lang.Thread.run(Thread.java:748)
[error] (Global / bloopInstall) sbt.internal.util.Init$RuntimeUndefined: References to undefined settings at runtime.
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-cassandra)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-service-locator)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-kafka)), Select(ConfigKey(test)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-service-locator)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-cassandra)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] setting(ScopedKey(Global,bloopInstall)) at LinePosition((bloop.integrations.sbt.PluginImplementation.globalSettings) SbtBloop.scala,56) referenced from ScopedKey(Scope(Select(ProjectRef(file:/opt/hello/,lagom-internal-meta-project-kafka)), Select(ConfigKey(compile)), Zero, Zero),bloopGenerate)
[error] Total time: 0 s, completed Apr 22, 2018 8:47:10 AM
```

### Output of sbt plugins in hello project
```
[/opt/hello]$ sbt plugins
[info] Loading settings from plugins.sbt ...
[info] Loading project definition from /opt/hello/project
[info] Loading settings from build.sbt ...
[info] Set current project to hello (in build file:/opt/hello/)
In file:/opt/hello/
        sbt.plugins.IvyPlugin: enabled in hello-api, hello-stream-api, hello-impl, hello-stream-impl, hello
        sbt.plugins.JvmPlugin: enabled in hello-api, hello-stream-api, hello-impl, hello-stream-impl, hello
        sbt.plugins.CorePlugin: enabled in hello-api, hello-stream-api, hello-impl, hello-stream-impl, hello
        sbt.plugins.JUnitXmlReportPlugin: enabled in hello-api, hello-stream-api, hello-impl, hello-stream-impl, hello
        sbt.plugins.Giter8TemplatePlugin: enabled in hello-api, hello-stream-api, hello-impl, hello-stream-impl, hello
        com.lightbend.lagom.sbt.Lagom: enabled in hello-impl, hello-stream-impl
        com.lightbend.lagom.sbt.LagomAkkaHttpServer: enabled in hello-impl, hello-stream-impl
        com.lightbend.lagom.sbt.LagomExternalProject
        com.lightbend.lagom.sbt.LagomJava
        com.lightbend.lagom.sbt.LagomLog4j2
        com.lightbend.lagom.sbt.LagomLogback: enabled in hello-impl, hello-stream-impl
        com.lightbend.lagom.sbt.LagomNettyServer
        com.lightbend.lagom.sbt.LagomPlay
        com.lightbend.lagom.sbt.LagomPlayJava
        com.lightbend.lagom.sbt.LagomPlayScala
        com.lightbend.lagom.sbt.LagomPlugin: enabled in hello-impl, hello-stream-impl
        com.lightbend.lagom.sbt.LagomReloadableService: enabled in hello-impl, hello-stream-impl
        com.lightbend.lagom.sbt.LagomScala: enabled in hello-impl, hello-stream-impl
        play.sbt.Play
        play.sbt.PlayAkkaHttp2Support
        play.sbt.PlayAkkaHttpServer
        play.sbt.PlayFilters
        play.sbt.PlayJava
        play.sbt.PlayLayoutPlugin
        play.sbt.PlayLogback
        play.sbt.PlayMinimalJava
        play.sbt.PlayNettyServer
        play.sbt.PlayScala
        play.sbt.PlayService
        play.sbt.routes.RoutesCompiler
        play.sbt.test.MediatorWorkaroundPlugin
        play.twirl.sbt.SbtTwirl
        com.typesafe.sbt.SbtNativePackager: enabled in hello-impl, hello-stream-impl
        com.typesafe.sbt.packager.archetypes.JavaAppPackaging: enabled in hello-impl, hello-stream-impl
        com.typesafe.sbt.packager.archetypes.JavaServerAppPackaging
        com.typesafe.sbt.packager.archetypes.jar.ClasspathJarPlugin
        com.typesafe.sbt.packager.archetypes.jar.LauncherJarPlugin
        com.typesafe.sbt.packager.archetypes.scripts.AshScriptPlugin
        com.typesafe.sbt.packager.archetypes.scripts.BashStartScriptPlugin: enabled in hello-impl, hello-stream-impl
        com.typesafe.sbt.packager.archetypes.scripts.BatStartScriptPlugin: enabled in hello-impl, hello-stream-impl
        com.typesafe.sbt.packager.archetypes.systemloader.SystemVPlugin
        com.typesafe.sbt.packager.archetypes.systemloader.SystemdPlugin
        com.typesafe.sbt.packager.archetypes.systemloader.SystemloaderPlugin
        com.typesafe.sbt.packager.archetypes.systemloader.UpstartPlugin
        com.typesafe.sbt.packager.debian.DebianDeployPlugin
        com.typesafe.sbt.packager.debian.DebianPlugin: enabled in hello-impl, hello-stream-impl
        com.typesafe.sbt.packager.debian.JDebPackaging
        com.typesafe.sbt.packager.docker.DockerPlugin: enabled in hello-impl, hello-stream-impl
        com.typesafe.sbt.packager.docker.DockerSpotifyClientPlugin
        com.typesafe.sbt.packager.jdkpackager.JDKPackagerDeployPlugin
        com.typesafe.sbt.packager.jdkpackager.JDKPackagerPlugin
        com.typesafe.sbt.packager.linux.LinuxPlugin: enabled in hello-impl, hello-stream-impl
        com.typesafe.sbt.packager.rpm.RpmDeployPlugin
        com.typesafe.sbt.packager.rpm.RpmPlugin: enabled in hello-impl, hello-stream-impl
        com.typesafe.sbt.packager.universal.UniversalDeployPlugin
        com.typesafe.sbt.packager.universal.UniversalPlugin: enabled in hello-impl, hello-stream-impl
        com.typesafe.sbt.packager.windows.WindowsDeployPlugin
        com.typesafe.sbt.packager.windows.WindowsPlugin: enabled in hello-impl, hello-stream-impl
        com.lightbend.sbt.javaagent.JavaAgent
        com.lightbend.sbt.javaagent.JavaAgentPackaging
        com.typesafe.sbt.web.SbtWeb
        com.typesafe.sbt.jse.SbtJsEngine
        com.typesafe.sbt.jse.SbtJsTask
        com.typesafe.sbteclipse.plugin.EclipsePlugin: enabled in hello-api, hello-stream-api, hello-impl, hello-stream-impl, hello
        bloop.integrations.sbt.SbtBloop: enabled in hello-api, hello-stream-api, hello-impl, hello-stream-impl, hello
```

### Output of sbt projects
```
[/opt/hello]$ sbt projects
[info] Loading settings from plugins.sbt ...
[info] Loading project definition from /opt/hello/project
[info] Loading settings from build.sbt ...
[info] Set current project to hello (in build file:/opt/hello/)
[info] In file:/opt/hello/
[info]   * hello
[info]     hello-api
[info]     hello-impl
[info]     hello-stream-api
[info]     hello-stream-impl
[info]     lagom-internal-meta-project-cassandra
[info]     lagom-internal-meta-project-kafka
[info]     lagom-internal-meta-project-service-locator
```

**Comment by jvican** [comment]
Thanks for the detailed reproduction @matthewpflueger. I'll give it a try soon and find out what's happening.

**Comment by jvican** [comment]
The issue here is that `LagomPlugin` adds projects of its own into the sbt build in the `onLoad` startup hook, and the newly added projects don't have the plugins that would be automatically added to every sbt project.
  
I worked around this issue in a more general way by changing the semantics of `bloopInstall`. `bloopInstall` will only be run on those projects that have the `BloopPlugin` enabled. Thanks for the high-quality ticket @matthewpflueger.

**Comment by jvican** [comment]
Fixed by #512.
